caffe2/mkl/operators/fully_connected_op.cc

```diff
@@ -0,0 +1,108 @@
+#include "caffe2/core/context.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/utils/mkl_utils.h"
+
+#ifdef CAFFE2_HAS_MKL_DNN
+
+namespace caffe2 {
+namespace mkl {
+
+template <typename T>
+class MKLFullyConnectedOp final : public MKLOperator<T> {
+ public:
+
+  MKLFullyConnectedOp(const OperatorDef& operator_def, Workspace* ws)
+      : MKLOperator<T>(operator_def, ws),
+        axis_(OperatorBase::GetSingleArgument<int32_t>("axis", 1)) {}
+  ~MKLFullyConnectedOp() {}
+
+  bool RunOnDevice() override {
+    auto& X = OperatorBase::Input<MKLMemory<float>>(INPUT);
+    auto& filter = OperatorBase::Input<MKLMemory<float>>(FILTER);
+    auto& bias = OperatorBase::Input<MKLMemory<float>>(BIAS);
+    MKLMemory<float>* Y = OperatorBase::Output<MKLMemory<float>>(0);
+        
+    CAFFE_ENFORCE(filter.ndim() == 2, filter.ndim());
+    CAFFE_ENFORCE(bias.ndim() == 1, bias.ndim());
+
+    if (cached_input_dims_ != X.dims() ||
```

Comment:
Alex, can you please explain what do we mean by layout here? AFAIK the MKL primitive/buffers/layouts has to be created newly only if the dimention changes.